modules/axtask/src/wait_queue.rs

```diff
@@ -135,50 +149,50 @@ impl WaitQueue {
     where
         F: Fn() -> bool,
     {
-        let curr = crate::current();
+        declare_current_waiter!(waiter);
         let deadline = axhal::time::wall_time() + dur;
         debug!(
             "task wait_timeout: {}, deadline={:?}",
-            curr.id_name(),
+            waiter.inner().id_name(),
             deadline
         );
-        crate::timers::set_alarm_wakeup(deadline, curr.clone());
+        crate::timers::set_alarm_wakeup(deadline, waiter.inner().clone());
 
         let mut timeout = true;
         loop {
-            let mut rq = current_run_queue::<NoPreemptIrqSave>();
             if axhal::time::wall_time() >= deadline {
                 break;
             }
-            let wq = self.queue.lock();
             if condition() {
                 timeout = false;
                 break;
             }
 
-            rq.blocked_resched(wq);
+            self.waiter_blocked(&waiter);
             // Preemption may occur here.
+            current_run_queue::<NoPreemptIrqSave>().reschedule();
         }
+
         // Always try to remove the task from the timer list.
-        self.cancel_events(curr, true);
+        self.cancel_events(&waiter, true);
         timeout
     }
 
-    /// Wakes up one task in the wait queue, usually the first one.
+    /// Wakes up one task in the wait list, usually the first one.
     ///
     /// If `resched` is true, the current task will be preempted when the
     /// preemption is enabled.
     pub fn notify_one(&self, resched: bool) -> bool {
-        let mut wq = self.queue.lock();
-        if let Some(task) = wq.pop_front() {
-            unblock_one_task(task, resched);
+        let mut wq = self.list.lock();
+        if let Some(waiter) = wq.pop_front() {
+            unblock_one_task(waiter.inner().clone(), resched);
```

Comment:
Add an `into_inner` method to the node type to avoid `Arc::clone`.

Reply:
In fact, we cannot use into_inner directly here, because there is a waiter reference in the blocked task context, and the waiter cannot be dropped directly.

Reply:
I used `into_inner` in `cancel_events` to ensure that the reference count is equal to 1 when the waiter is destroyed.